File: Code_Examples/Chapter_02/Src/C/area.omp.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <omp.h>

# define NPOINTS 4000
# define MAXITER 4000


struct complex{
  double real;
  double imag;
};

int main(){
  int i, j, iter, numoutside = 0;
  int num_threads;
  double area, error, ztemp;
  struct complex z, c;

  double start, end, cpu_time_used;
     
  start = omp_get_wtime();
/*
 *   
 *
 *     Outer loops run over npoints, initialise z=c
 *
 *     Inner loop has the iteration z=z*z+c, and threshold test
 */
#pragma omp parallel for default(shared) private(i,j,c,z,ztemp,iter) reduction(+:numoutside) collapse(2)
  for (i=0; i<NPOINTS; i++) {
    for (j=0; j<NPOINTS; j++) {
      c.real = -2.0+2.5*(double)(i)/(double)(NPOINTS)+1.0e-7;
      c.imag = 1.125*(double)(j)/(double)(NPOINTS)+1.0e-7;
      z=c;
      for (iter=0; iter<MAXITER; iter++){
	ztemp=(z.real*z.real)-(z.imag*z.imag)+c.real;
	z.imag=z.real*z.imag*2+c.imag; 
	z.real=ztemp; 
	if ((z.real*z.real+z.imag*z.imag)>4.0e0) {
	  numoutside++;
          break; 
	}
      }
    }
  }



  end = omp_get_wtime();
  cpu_time_used = ((double) (end - start));

#pragma omp parallel
{
  num_threads = omp_get_num_threads();
}


/*
 *  Calculate area and error and output the results
 */

   area=2.0*2.5*1.125*(double)(NPOINTS*NPOINTS-numoutside)/(double)(NPOINTS*NPOINTS);
   error=area/(double)NPOINTS;

   printf("Area of Mandlebrot set = %12.8f +/- %12.8f\n",area,error);
   printf("Time taken for calculation: %f on %d threads\n",cpu_time_used,num_threads);    
}

File: Code_Examples/Chapter_02/Src/Fortran/area.omp.f90
program area_mandelbrot

  use omp_lib

  implicit none

  integer, parameter :: sp = kind(1.0)   
  integer, parameter :: dp = kind(1.0d0)
  integer :: i, j, iter, numoutside, numthreads
  integer, parameter :: npoints = 2000, maxiter = 2000
  real (kind=dp) :: area, error
  complex (kind=dp) :: c , z
  real :: starttime, endtime
!
! Calculate area of mandelbrot set
!
!    Outer loops runs over npoints, initialize z=c
!
!    Inner loop has the iteration z=z*z+c, and threshold test
!

  numoutside = 0 

  starttime = omp_get_wtime()

  !$omp parallel do reduction(+:numoutside) private(i,j,iter,c,z) collapse(2)
  do i = 0,npoints-1
     do j= 0,npoints-1 
        c = cmplx(-2.0+(2.5*i)/npoints + 1.0d-07,(1.125*j)/npoints + 1.0d-07)
        z = c
        iter = 0
        do while (iter < maxiter) 
           z = z*z + c 
           iter = iter + 1
           if (real(z)*real(z)+aimag(z)*aimag(z) > 4) then
              numoutside = numoutside + 1 
              exit
           endif
        end do 
     end do
  end do

  endtime = omp_get_wtime()

  !$omp parallel
   numthreads = omp_get_num_threads()
  !$omp end parallel

!
! Output results
!
  area = 2.0*2.5*1.125 * real(npoints*npoints-numoutside)/real(npoints*npoints)
  error = area/real(npoints)
  print *, "Area of Mandelbrot set = ",area," +/- ",error
  print *, "Time taken for calculation: ",endtime-starttime
!
  stop
end program area_mandelbrot
